cpp/src/stream_compaction/distinct.cu

```diff
@@ -80,21 +81,62 @@ std::unique_ptr<table> distinct(table_view const& input,
   // insert distinct indices into the map.
   key_map.insert(iter, iter + num_rows, hash_key, key_equal, stream.value());
 
+#if 0
   auto const output_size{key_map.get_size()};
   auto distinct_indices = cudf::make_numeric_column(
     data_type{type_id::INT32}, output_size, mask_state::UNALLOCATED, stream, mr);
   // write distinct indices to a numeric column
   key_map.retrieve_all(distinct_indices->mutable_view().begin<cudf::size_type>(),
                        thrust::make_discard_iterator(),
                        stream.value());
-
   // run gather operation to establish new order
   return detail::gather(input,
                         distinct_indices->view(),
                         out_of_bounds_policy::DONT_CHECK,
                         detail::negative_index_policy::NOT_ALLOWED,
                         stream,
                         mr);
+#else
+
+  auto const key_size{keys_view.num_rows()};
+  auto distinct_indices = rmm::device_uvector<size_type>(key_size, stream);
+
+  // Fill `key_size` if keep_first
+  // Fill `INT_MIN` if keep_last
+  thrust::uninitialized_fill(rmm::exec_policy(stream),
+                             distinct_indices.begin(),
+                             distinct_indices.begin() + key_size,
+                             key_size);
+
+  auto const d_map = key_map.get_device_view();
+  thrust::for_each(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(key_size),
+    [distinct_indices = distinct_indices.begin(), d_map, hash_key, key_equal] __device__(auto key) {
+      // iter should always be valid, because all keys have been inserted.
+      // Here idx is the index of the unique elements that has been inserted into the map.
+      // As such, `find` calling for all duplicate keys will return the same idx value.
+      auto const idx =
+        d_map.find(key, hash_key, key_equal)->second.load(cuda::std::memory_order_relaxed);
+
+      // Store the minimum index of all keys that are equal.
+      atomicMin(&distinct_indices[idx], key);
```

Comment:
We should avoid introducing new code dependent on legacy `atomic*` constructors and prefer use of `cuda::atomic/cuda::atomic_ref`. 
```suggestion
      // Store the minimum index of all keys that are equal.
      cuda::atomic_ref<size_type, thread_scope_device> min_index{distinct_indices[idx]};
      min_index.fetch_min(key, memory_order_relaxed);
```

Reply:
The `fetch_min` and `fetch_max` are just for reading, not for writing (although they can do writing but it's not guaranteed).
Ref: https://github.com/NVIDIA/libcudacxx/blob/main/docs/extended_api/synchronization_primitives/atomic.md#atomic-extrema-operations

I just tested that. `fetch_min` replaces the old value with the smaller value but `fetch_max` does nothing.

Reply:
It seems that there is a bug: https://github.com/NVIDIA/libcudacxx/issues/279

Reply:
So in the meantime, I'll just use `atomicMin`/`atomicMax` until that is fixed.

Reply:
> The fetch_min and fetch_max are just for reading, not for writing (although they can do writing but it's not guaranteed).

That is not correct. `fetch_min/fetch_max` are read-modify-write operations that will update the existing value. 

> It seems that there is a bug: https://github.com/NVIDIA/libcudacxx/issues/279

That bug is only relevant to floating point values. It should work fine for integral types. (Do let me know if that is not the case).

Reply:
The bug is also relevant for integer type. I have reported my test in the link. Here is the post again:
----
I also got incorrect result with negative values:
 * Fill an int array with `std::numeric_limits<size_type>::min()`
 * Feed a sequence of `[0, 1, 2, ...]` into a function that:
 ```
__device__ void f(int idx) {
    auto const atomic_idx =
          cuda::atomic_ref<size_type, cuda::thread_scope::thread_scope_device>{d_output[idx]};
    atomic_idx.fetch_max(idx, cuda::std::memory_order_relaxed);
}
```
And the result is `[-2147483648,-2147483648,......]`.

Reply:
I can confirm that I can generate the correct result for `cudf::distinct` with `atomicMax` but not with `fetch_max`.

Reply:
> That is not correct. fetch_min/fetch_max are read-modify-write operations that will update the existing value.

If so, the docs are incorrect (https://github.com/NVIDIA/libcudacxx/blob/main/docs/extended_api/synchronization_primitives/atomic.md#atomic-extrema-operations) and should be updated.

Reply:
> If so, the docs are incorrect

How so? 

> Atomically find the minimum of the stored value and a provided value. (member function)

https://nvidia.github.io/libcudacxx/extended_api/synchronization_primitives/atomic/fetch_min.html#cudaatomicfetch_min

https://nvidia.github.io/libcudacxx/extended_api/synchronization_primitives/atomic.html#atomic-extrema-operations

Reply:
Oh I state again: the docs are correct but not enough. For example, for `fetch_min`: "Atomically find the minimum of the stored value and a provided value.". That tells me it just reads the value but does not write anything.